addons/website_slides_survey/models/survey_user.py

```diff
@@ -32,19 +34,39 @@ def _check_for_failed_attempt(self):
 
         The purpose is to have a 'certification flow' where the user can re-purchase the
         certification when they have failed it."""
-        for record in self:
-            if record.state == 'done' and not record.quizz_passed and record.slide_partner_id:
-                if not record.survey_id._has_attempts_left(record.partner_id, record.email, record.invite_token):
-                    # The mail needs to be sent BEFORE deleting slide and channel relations because we use them in the layout
-                    self.env.ref('website_slides_survey.mail_template_user_input_certification_failed').send_mail(record.id, notif_layout="mail.mail_notification_light")
-
-                    channel_id = record.slide_partner_id.channel_id.id
-                    self.env['slide.slide.partner'].sudo().search([
-                        ('partner_id', '=', record.partner_id.id),
-                        ('slide_id', 'in', record.slide_partner_id.channel_id.slide_ids.ids)
-                    ]).unlink()
-
-                    self.env['slide.channel.partner'].sudo().search([
-                        ('partner_id', '=', record.partner_id.id),
-                        ('channel_id', '=', channel_id)
-                    ]).unlink()
+
+        if len(self) > 0:
+            user_inputs = self.search([
+                ('id', 'in', self.ids),
+                ('state', '=', 'done'),
+                ('quizz_passed', '=', False),
+                ('slide_partner_id', '!=', False)
+            ])
+
+            if user_inputs:
+                removed_slide_partner_domain = []
+                removed_channel_partner_domain = []
+                for user_input in user_inputs:
+                    if user_input.survey_id._has_attempts_left(user_input.partner_id, user_input.email, user_input.invite_token):
+                        # skip if user still has attempts left
+                        continue
+
+                    self.env.ref('website_slides_survey.mail_template_user_input_certification_failed').send_mail(
+                        user_input.id, notif_layout="mail.mail_notification_light"
+                    )
+                    slide_partner_id = user_input.slide_partner_id
+                    removed_slide_partner_domain = expression.OR([
```

Comment:
My only fear is to have a too big domain at the end. But, I think making a write with 'state' value is rare and normally only append once on only one record, right ?

Reply:
Making a write with 'state' might happen multiple times because it also has a 'skip' value.
However, writing it to 'done' will happen only once per record. So it should be fine.